operators/hack/deployer/gke.go

```diff
@@ -0,0 +1,263 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+)
+
+const (
+	GkeDriverId                   = "gke"
+	GkeServiceAccountKeyVaultName = "secret/cloud-team/cloud-ci/ci-gcp-k8s-operator"
+)
+
+func init() {
+	drivers[GkeDriverId] = &GkeDriverFactory{}
+}
+
+type GkeDriverFactory struct {
+}
+
+type GkeDriver struct {
+	plan Plan
+}
+
+func (gdf *GkeDriverFactory) Create(plan Plan) (Driver, error) {
+	return &GkeDriver{plan: plan}, nil
+}
+
+func (d *GkeDriver) Execute() error {
+	if err := d.auth(); err != nil {
+		return err
+	}
+
+	exists, err := d.clusterExists()
+	if err != nil {
+		return err
+	}
+
+	switch d.plan.Operation {
+	case "delete":
+		if exists {
+			err = d.delete()
+		} else {
+			log.Printf("not deleting as cluster doesn't exist")
+		}
+	case "create":
+		if exists {
+			log.Printf("not creating as cluster exists")
+		} else {
+			if err := d.configSsh(); err != nil {
+				return err
+			}
+
+			if err := d.create(); err != nil {
+				return err
+			}
+
+			if err := d.bindRoles(); err != nil {
+				return err
+			}
+		}
+
+		if err := d.getCredentials(); err != nil {
+			return err
+		}
+
+		if d.plan.VmMapMax {
+			if err := d.setMaxMapCount(); err != nil {
+				return err
+			}
+		}
+
+		if err := d.configureDocker(); err != nil {
+			return err
+		}
+	default:
+		err = fmt.Errorf("unknown operation %s", d.plan.Operation)
+	}
+
+	return err
+}
+
+func (d *GkeDriver) auth() error {
+	if d.plan.ServiceAccount {
+		return d.serviceAuth()
+	} else {
+		return d.userAuth()
+	}
+}
+
+func (d *GkeDriver) serviceAuth() error {
+	log.Println("Authenticating as service account...")
+
+	serviceAccountKey, err := ReadVault(d.plan.VaultAddress, d.plan.VaultRoleId, d.plan.VaultSecretId, GkeServiceAccountKeyVaultName)
+	if err != nil {
+		return err
+	}
+
+	file, err := ioutil.TempFile(".", "gke_service_account_key-")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(file.Name())
+
+	if err := ioutil.WriteFile(file.Name(), []byte(serviceAccountKey), 0644); err != nil {
+		return err
+	}
+
+	return NewCommand("gcloud auth activate-service-account --key-file={{.FileName}}").
```

Comment:
Do we need to shell out here instead of using the official SDK? Or is that an improvement for later? It looks like they provide funcs for the CLI calls I spot checked here

Reply:
In general I'd strongly prefer sdk, but right now I think we should focus on moving this forward to enable more providers. Moving later to the sdk should be a fairly localized change.